### DISCUSSION TITLE
Handle situation where there are fewer operations than machines in jobshop_sat.cc

### DISCUSSION THREAD

**Discussion by Lederilon** [google/or-tools#2699] [CP-SAT questions] [answered]
For example, we have 4 operations and 6 machines in this example - 
https://github.com/google/or-tools/blob/stable/examples/cpp/jobshop_sat.cc

So for at least two machines, circuit constraint will be infeasible. 
For that machine, all machine_to_tasks[m][i].presence would be false and no path from 0 -> 0 exists

**Comment by lperron** [comment, marked as answer]
All nodes must have one 1 and 1 out arc in the circuit constraint.

So for each task, add a self looping arc with the negation of the performed literal of the task as arc literal.
Add a looping node on the start/end node (0) with the new boolean variable indicating if the resource is used or not.
For all nodes, add an implication task is performed implies machine is used.

**Reply by Lederilon** [reply]
So it would be a new arc from 0 to 0 with literal machineUsed.Not() and implication(task resence, machineUsed)?
Can you add your improvements to jobshop_sat.cc example?

**Reply by lperron** [reply]
Indeed. Implemented.
